File: src/main/java/com/library/observers/NotificationService.java
package com.library.observers;

import java.util.ArrayList;
import java.util.List;

/**
 * Service de notification qui implémente le pattern Observable
 * Gère les abonnés et diffuse les notifications
 */
public class NotificationService {
    private List<NotificationObserver> observers = new ArrayList<>();

    public void subscribe(NotificationObserver observer) {
        if (!observers.contains(observer)) {
            observers.add(observer);
        }
    }

    public void unsubscribe(NotificationObserver observer) {
        observers.remove(observer);
    }

    public void notifyObservers(String message) {
        for (NotificationObserver observer : observers) {
            observer.update(message);
        }
    }
}

File: js/src/observers/NotificationService.js
/**
 * Service de notification - Pattern Observer
 * Gère les abonnés et diffuse les notifications
 */
class NotificationService {
    constructor() {
        this.observers = [];
    }

    /**
     * S'abonner aux notifications
     * @param {NotificationObserver} observer - L'observateur à ajouter
     */
    subscribe(observer) {
        if (!this.observers.includes(observer)) {
            this.observers.push(observer);
        }
    }

    /**
     * Se désabonner aux notifications
     * @param {NotificationObserver} observer - L'observateur à retirer
     */
    unsubscribe(observer) {
        const index = this.observers.indexOf(observer);
        if (index > -1) {
            this.observers.splice(index, 1);
        }
    }

    /**
     * Notifier tous les observateurs
     * @param {string} message - Le message à envoyer
     */
    notifyObservers(message) {
        this.observers.forEach(observer => {
            observer.update(message);
        });
    }
}

module.exports = NotificationService;
